src/Features/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicDocumentPullDiagnosticsHandler_IOnInitialized.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.Public;
+
+using DocumentDiagnosticReport = SumType<RelatedFullDocumentDiagnosticReport, RelatedUnchangedDocumentDiagnosticReport>;
+
+// A document diagnostic partial report is defined as having the first literal send = DocumentDiagnosticReport (aka changed / unchanged) followed
+// by n DocumentDiagnosticPartialResult literals.
+// See https://github.com/microsoft/vscode-languageserver-node/blob/main/protocol/src/common/proposed.diagnostics.md#textDocument_diagnostic
+using DocumentDiagnosticPartialReport = SumType<RelatedFullDocumentDiagnosticReport, RelatedUnchangedDocumentDiagnosticReport, DocumentDiagnosticReportPartialResult>;
+
+internal sealed partial class PublicDocumentPullDiagnosticsHandler : AbstractDocumentPullDiagnosticHandler<DocumentDiagnosticParams, DocumentDiagnosticPartialReport, DocumentDiagnosticReport?>, IOnInitialized
+{
+    public async Task OnInitializedAsync(ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+    {
+        if (clientCapabilities?.TextDocument?.Diagnostic?.DynamicRegistration is true)
```

Comment:
Should we also check if FSA is on?  No point in asking the client to send us diagnostic requests that we're never going to return anything from.  We'd have to react to changes in the option as well (think there should be a global option change event), but it would save us all the queuing time, serialization / deserialization, etc.  Especially since these requests happen on typing.

Fine with that in a separate PR though

Reply:
> We'd have to react to changes in the option as well 

I wasn't sure about how exactly to go about hooking this up. I'll take a quick look and if I am able to will do these changes in this PR. Otherwise, will file a tracking issue for follow-up.

Reply:
So, the background analysis scope option is defined per-language. I'll figure out how to fetch this for all supported languages in this Initialize method.

Reply:
@dibarbet - done with https://github.com/dotnet/roslyn/pull/69541/commits/0121aa0a2d238c1140819ad82bbc395fc3ab0661